platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/collections/AbstractLongList.java

```diff
@@ -0,0 +1,545 @@
+    /** The number for bytes to read for file header, v2 */
+    protected static final int FILE_HEADER_SIZE_V2 = VERSION_METADATA_SIZE + FORMAT_METADATA_SIZE_V2;
+    /** File header size for the latest format */
+    protected final int currentFileHeaderSize;
+
+    /**
+     * The number of longs to store in each allocated buffer. Must be a positive integer. If the
+     * value is small, then we will end up allocating a very large number of buffers. If the value
+     * is large, then we will waste a lot of memory in the unfilled buffer.
+     */
+    protected final int numLongsPerChunk;
+    /** Size in bytes for each memory chunk to allocate */
+    protected final int memoryChunkSize;
+    /** The number of longs that this list would contain if it was not optimized by {@link AbstractLongList#updateMinValidIndex}.
+     * Practically speaking, it defines the list's right boundary. */
+    protected final AtomicLong size = new AtomicLong(0);
+    /**
+     * The maximum number of longs to ever store in this data structure. This is used as a safety
+     * measure to make sure no bug causes an out of memory issue by causing us to allocate more
+     * buffers than the system can handle.
+     */
+    protected final long maxLongs;
+
+    /** Min valid index of the list. All the indices to the left of this index have {@code IMPERMISSIBLE_VALUE}-s */
+    protected final AtomicLong minValidIndex = new AtomicLong(0);
+
+    /** Atomic reference array of our memory chunks */
+    protected final AtomicReferenceArray<C> chunkList;
+
+    /**
+     * A length of a buffer that is reserved to remain intact after memory optimization that is
+     * happening in {@link LongListOffHeap#updateMinValidIndex}
+     */
+    protected final long reservedBufferLength;
+
+    /**
+     * Construct a new LongList with the specified number of longs per chunk and maximum number of
+     * longs.
+     *
+     * @param numLongsPerChunk number of longs to store in each chunk of memory allocated
+     * @param maxLongs the maximum number of longs permissible for this LongList
+     * @param reservedBufferLength reserved buffer length that the list should have before minimal index in the list
+     */
+    protected AbstractLongList(final int numLongsPerChunk, final long maxLongs, final long reservedBufferLength) {
+        if (maxLongs < 0) {
+            throw new IllegalArgumentException("The maximum number of longs must be non-negative, not " + maxLongs);
+        }
+        if (numLongsPerChunk > MAX_NUM_LONGS_PER_CHUNK) {
+            throw new IllegalArgumentException(
+                    "Cannot store " + numLongsPerChunk + " per chunk (max is " + MAX_NUM_LONGS_PER_CHUNK + ")");
+        }
+        this.maxLongs = maxLongs;
+        this.numLongsPerChunk = numLongsPerChunk;
+        final int chunkNum = calculateNumberOfChunks(maxLongs);
+        if (chunkNum > MAX_NUM_CHUNKS) {
+            throw new IllegalArgumentException(String.format(
```

Comment:
This could be cleaner with a constant.
```
Constant:
private static final String MAX_CHUNKS_EXCEEDED_MSG =
        "The maximum number of memory chunks should not exceed %s."
        + " Either increase numLongsPerChunk or decrease maxLongs";

Modified Line:
throw new IllegalArgumentException(MAX_CHUNKS_EXCEEDED_MSG.formatted(MAX_NUM_CHUNKS));
```

Reply:
Agreed.

Reply:
?